Userland/Libraries/LibIMAP/Client.cpp

```diff
@@ -0,0 +1,459 @@
+{
+    Vector<String> args;
+    if (charset.has_value()) {
+        args.append("CHARSET ");
+        args.append(charset.value());
+    }
+    for (const auto& item : keys) {
+        args.append(item.serialize());
+    }
+    auto command = Command { uid ? CommandType::UIDSearch : CommandType::Search, m_current_command, args };
+    return cast_promise<SolidResponse>(send_command(move(command)));
+}
+
+RefPtr<Promise<Optional<ContinueRequest>>> Client::idle()
+{
+    auto promise = send_simple_command(CommandType::Idle);
+    return cast_promise<ContinueRequest>(promise);
+}
+RefPtr<Promise<Optional<SolidResponse>>> Client::finish_idle()
+{
+    auto promise = Promise<Optional<Response>>::construct();
+    m_pending_promises.append(promise);
+    send_raw("DONE");
+    m_expecting_response = true;
+    return cast_promise<SolidResponse>(promise);
+}
+
+RefPtr<Promise<Optional<SolidResponse>>> Client::status(StringView mailbox, const Vector<StatusItemType>& types)
+{
+    Vector<String> args;
+    for (auto type : types) {
+        switch (type) {
+        case StatusItemType::Recent:
+            args.append("RECENT");
+            break;
+        case StatusItemType::UIDNext:
+            args.append("UIDNEXT");
+            break;
+        case StatusItemType::UIDValidity:
+            args.append("UIDVALIDITY");
+            break;
+        case StatusItemType::Unseen:
+            args.append("UNSEEN");
+            break;
+        case StatusItemType::Messages:
+            args.append("MESSAGES");
+            break;
+        }
+    }
+    StringBuilder types_list;
+    types_list.append('(');
+    types_list.join(" ", args);
+    types_list.append(')');
+    auto command = Command { CommandType::Status, m_current_command, { mailbox, types_list.build() } };
+    return cast_promise<SolidResponse>(send_command(move(command)));
+}
+
+RefPtr<Promise<Optional<SolidResponse>>> Client::append(StringView mailbox, Message&& message, Optional<Vector<String>> flags, Optional<Core::DateTime> date_time)
+{
+    Vector<String> args = { mailbox };
+    if (flags.has_value()) {
+        StringBuilder flags_sb;
+        flags_sb.append('(');
+        flags_sb.join(" ", flags.value());
+        flags_sb.append(')');
+        args.append(flags_sb.build());
+    }
+    if (date_time.has_value())
+        args.append(date_time.value().to_string("\"%d-%b-%Y %H:%M:%S +0000\""));
+
+    args.append(String::formatted("{{{}}}", message.data.length()));
+
+    auto continue_req = send_command(Command { CommandType::Append, m_current_command, args });
+
+    RefPtr<Promise<Optional<Response>>> response_promise = Promise<Optional<Response>>::construct();
+    m_pending_promises.append(response_promise);
+
+    continue_req->callback = [this, message2 { move(message) }] {
+        send_raw(message2.data);
+        m_expecting_response = true;
+    };
+
+    return cast_promise<SolidResponse>(response_promise);
```

Comment:
The typical C++-ish name would be `promise_cast`, but I'm not sure what this even does?
Maybe it could instead be a method on Promise?

Reply:
It's a little utility function that essentially converts a `Promise<Variant<A, B>>` into either a `Promise<A>` or a `Promise<B>`. `send_command` returns a `Promise<Variant>` since it doesn't know what type of response we expect. However, for some commands we do know what type of response the server will reply with, so I use this helper to convert the response returned by `send_command`. Don't think it would make sense to make it a method on `Promise`.

Reply:
Ah, it's a library specific thing, got it.